File: inference_framework/inference_framework/log_processor.py
from typing import List, Dict, Any
import logging
from inference_framework.event_processor import EventProcessor
from inference_framework.feature_store import FeatureStore
from inference_framework.model_loader import ModelLoader
from inference_framework.output_event import OutputEvent

logger = logging.getLogger(__name__)

class LogProcessor:
    """Processes batches of logs: to DF, fetch features, run the SINGLE loaded model."""
    
    def __init__(self, feature_store: FeatureStore, model_loader: ModelLoader):
        self.feature_store = feature_store
        self.model_loader = model_loader
        self.event_processor = EventProcessor()
        # Get the single model (enforced by ModelLoader)
        self.model = model_loader.get_model()
        if not self.model:
            raise ValueError("No model loaded by ModelLoader")
    
    def process_batch(self, validated_batch: List[Dict[str, Any]]) -> List[OutputEvent]:
        """Process validated batch: DF + features + inference via the single model.
        Returns List[OutputEvent] for Kafka producer."""
        if not validated_batch:
            return []
        
        # Convert to Polars DF
        df = self.event_processor.process_batch(validated_batch)
        if df.is_empty():
            return []
        
        # Get unique user_ids
        user_ids = df['user_id'].unique().to_list() if 'user_id' in df.columns else []
        features = self.feature_store.get_features_for_users(user_ids)
        
        # Run the single model (no loop needed)
        all_events: List[OutputEvent] = []
        try:
            model_name = self.model_loader.get_model_name()
            model_events = self.model.process(df, features)
            all_events.extend(model_events)
        except Exception as e:
            logger.error(f"Error in model {model_name}: {str(e)}")
        
        logger.info(f"Generated {len(all_events)} inference events from batch using model: {model_name}")
        return all_events

File: inference_framework-java/src/main/java/com/inferenceframework/processing/LogProcessor.java
package com.inferenceframework.processing;  // processing module

import com.inferenceframework.event.OutputEvent;  // from event module
import com.inferenceframework.model.ModelInterface;
import com.inferenceframework.model.ModelLoader;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Processes batches of logs: fetch features, run the SINGLE loaded model.
 * Adapted: uses EventProcessor to List<Map> (no Polars), then model inference.
 * Thread-safe: stateless per-batch, shared deps (FeatureStore, models) are read-only/immutable after init.
 */
public class LogProcessor {
    private static final Logger logger = LoggerFactory.getLogger(LogProcessor.class);

    private final FeatureStore featureStore;
    private final ModelLoader modelLoader;
    private final EventProcessor eventProcessor;
    // Get the single model (enforced by ModelLoader)
    private final ModelInterface model;

    /**
     * Constructor.
     */
    public LogProcessor(FeatureStore featureStore, ModelLoader modelLoader) {
        this.featureStore = featureStore;
        this.modelLoader = modelLoader;
        this.eventProcessor = new EventProcessor();
        // Get the single model
        this.model = modelLoader.getModel();
        if (this.model == null) {
            throw new IllegalArgumentException("No model loaded by ModelLoader");
        }
    }

    /**
     * Process validated batch: to processable format + features + inference via the single model.
     * Returns List<OutputEvent> for Kafka producer.
     */
    public List<OutputEvent> processBatch(List<Map<String, Object>> validatedBatch) {
        if (validatedBatch == null || validatedBatch.isEmpty()) {
            return new ArrayList<>();
        }

        // Convert/process batch (returns List<Map> in Java)
        List<Map<String, Object>> processed = eventProcessor.processBatch(validatedBatch);
        if (processed.isEmpty()) {
            return new ArrayList<>();
        }

        // Get unique user_ids using helper
        Set<String> userIds = eventProcessor.extractUserIds(processed);
        Map<String, Map<String, Object>> features = featureStore.getFeaturesForUsers(new ArrayList<>(userIds));

        // Run the single model (no loop; model handles batch)
        List<OutputEvent> allEvents = new ArrayList<>();
        try {
            String modelName = modelLoader.getModelName();
            List<OutputEvent> modelEvents = model.process(processed, features);
            allEvents.addAll(modelEvents);
            logger.info("Generated {} inference events from batch using model: {}", allEvents.size(), modelName);
        } catch (Exception e) {
            String modelName = modelLoader.getModelName();
            logger.error("Error in model {}", modelName, e);
        }

        return allEvents;
    }
}